src/pip/_internal/operations/prepare.py

```diff
@@ -415,21 +354,25 @@ def prepare_linked_requirement(
         with indent_log():
             # Since source_dir is only set for editable requirements.
             assert req.source_dir is None
-            req.ensure_has_source_dir(self.build_dir, autodelete_unpacked)
-            # If a checkout exists, it's unwise to keep going.  version
-            # inconsistencies are logged later, but do not fail the
-            # installation.
-            # FIXME: this won't upgrade when there's an existing
-            # package unpacked in `req.source_dir`
-            if os.path.exists(os.path.join(req.source_dir, 'setup.py')):
-                raise PreviousBuildDirError(
-                    "pip can't proceed with requirements '{}' due to a"
-                    " pre-existing build directory ({}). This is "
-                    "likely due to a previous installation that failed"
-                    ". pip is being responsible and not assuming it "
-                    "can delete this. Please delete it and try again."
-                    .format(req, req.source_dir)
-                )
+            if link.is_existing_dir():
```

Comment:
Can we move this conditional into `ensure_has_source_dir` instead?

Reply:
Hm maybe, but then we'd need to special case `link.is_existing_dir` to avoid raising an exception a couple of lines below?

Actually it's probably worth doing some refactoring in the way we handle this source_dir. For instance why is it necessary to pass it as a constructor argument to `InstallRequirement`? Is it not redundant with `link`? And pushing this thought line forward, maybe it is enough to pass `req` and let `InstallRequirement` deduce `link` by itself? There is something bothering me around these methods since I starting trying to understand #8005 (which I still don't grok :).

Anyway, all this to say that I'd leave this to followup PRs.

Reply:
Fair enough. :)

FWIW, `link = req.link` which is what made me think of this but yea, this entire area of directory management could use a little cleanup -- which is why I had deferred tackling this issue personally. :P